File: go/strain/strain.go
package strain

type Ints []int
type Lists [][]int
type Strings []string

func (input Ints) Keep(predicate func(int) bool) Ints {
	var output Ints
	for _, v := range input {
		if predicate(v) {
			output = append(output, v)
		}

	}
	return output
}

//    (Ints) Discard(func(int) bool) Ints

func (input Ints) Discard(predicate func(int) bool) Ints {
	var output Ints
	for _, v := range input {
		if !predicate(v) {
			output = append(output, v)
		}

	}
	return output
}

func (input Lists) Keep(predicate func([]int) bool) Lists {
	var output Lists
	for _, v := range input {
		if predicate(v) {
			output = append(output, v)
		}
	}
	return output
}

func (input Strings) Keep(predicate func(string) bool) Strings {
	var output Strings
	for _, v := range input {
		if predicate(v) {
			output = append(output, v)
		}

	}
	return output
}

File: typescript/strain/strain.ts
export const keep = <T>(inputList: T[], predicate: (item: T) => boolean): T[] => {
  let output: T[] = []
  for (let listItem of inputList) {
    if (predicate(listItem)) {
      output.push(listItem)
    }
  }
  return output
}

export const discard = <T>(inputList: T[], predicate: (item: T) => boolean): T[] => {
  return keep(inputList, function (item: T) { return !predicate(item) })
}
